### PR TITLE
Lapack banded matrix wrappers

### PR DESCRIPTION
**Pull request by drwells**
I am working on some problems that would benefit from a very accurate banded matrix solver. LAPACK has good support for this.

This is a preliminary draft of a wrapper class for a LAPACK banded matrix: ~~so far, I have implemented the correct indexing and matrix-vector multiplication. I wanted to put this up as WIP since a lot of other people are working on LAPACK wrappers.~~ Update: see the checklist

On the TODO list:
- [x] storage scheme
- [x] test constructors
- [x] implement copy constructors for `FullMatrix`, `SparsityPattern`, and `SparseMatrix`
- [x] enable copy constructor for `LAPACKFullMatrix` (needs #5959)
- [x] wrap linear solvers (e.g., `gbtrf2`, `gbsvx`, `gbbrd`)
- [ ] support for symmetric calculations/symmetric mode
- [x] `vmult` and friends
- [ ] wrap some matrix multipliers
- [ ] changelog entry
- [ ] reimplement the tridiagonal matrix as a special case (i.e., use Thomas' algorithm if `n_subdiagonals = n_superdiagonals= 1`
- [ ] contiguous iterator implementation (needs #5959)
- [ ] diagonal iterator implementation (needs #5959)

### CODE REVIEW

File: include/deal.II/lac/lapack_banded_matrix.h [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+   */
+  typedef std::make_unsigned<types::blas_int>::type size_type;
+
+  /**
+   * Default constructor. Creates a size zero empty matrix.
+   */
+  LAPACKBandedMatrix() = default;
+
+  /**
+   * Constructor. Creates a logically <code>n_rows</code> by
```
Comment (davydden) [inline include/deal.II/lac/lapack_banded_matrix.h]: use `@p n_rows` and same for others?

Comment (drwells) [reply include/deal.II/lac/lapack_banded_matrix.h]: Fixed.

File: include/deal.II/lac/lapack_banded_matrix.h [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+DEAL_II_NAMESPACE_OPEN
+
+// forward declarations
+template <typename Number> class Vector;
+
+
+/**
+ * Class encapsulating scalars stored by LAPACKBandedMatrix.
+ */
+class LAPACKBandedMatrixData
```
[the remark this answers is not in the document]
Comment (drwells) [reply include/deal.II/lac/lapack_banded_matrix.h]: Probably not, but it makes copying and swapping easier.

I'll take another look at this class when I'm closer to the end to see if it actually helps.

**Comment by drwells** [comment]
While this is mostly done I won't have time to finish until the end of the year, so lets close for now.

**Comment by masterleinad** [comment]
I don't see a problem with having pull requests open that are mostly done and we want to have in the library even if it takes some for us to finish them. At least, people can see what we are wanting to do.

**Comment by drwells** [comment]
I know, but I keep seeing this on my list open PRs and its distracting to me. I have to close it for my own sake.

**Pull request closed without merging**